rflx/model.py

```diff
@@ -713,68 +721,134 @@ def __check_attributes(self, expression: Expr, location: Location = None) -> Non
+                if not first <= element <= last:
+                    self.error.append(
+                        f"aggregate element out of range {first} .. {last}",
+                        Subsystem.MODEL,
+                        Severity.ERROR,
+                        element.location,
+                    )
+
+        def check_enumeration(relation: Relation, left: Enumeration, right: Enumeration) -> None:
+            if left != right:
+                self.error.append(
+                    "comparison of incompatible enumeration literals",
+                    Subsystem.MODEL,
+                    Severity.ERROR,
+                    relation.location,
+                )
+                self.error.append(
+                    f'of type "{left.identifier}"', Subsystem.MODEL, Severity.INFO, left.location,
+                )
+                self.error.append(
+                    f'and type "{right.identifier}"',
+                    Subsystem.MODEL,
+                    Severity.INFO,
+                    right.location,
+                )
+
+        def relation_error(relation: Relation, left: TypeExpr, right: TypeExpr) -> None:
+            self.error.append(
+                f'invalid relation "{relation.symbol}" between {left.__class__.__name__} '
+                f"and {right.__class__.__name__}",
+                Subsystem.MODEL,
+                Severity.ERROR,
+                relation.location,
+            )
+
+        for relation in expression.findall(lambda x: isinstance(x, Relation)):
+            assert isinstance(relation, Relation)
+            left, right = self.__resolve_types(
+                relation.left.simplified(), relation.right.simplified(), literals
+            )
+            if isinstance(relation, (Less, LessEqual, Greater, GreaterEqual)):
+                if (isinstance(left, Aggregate) and isinstance(right, Type)) or (
+                    isinstance(right, Aggregate) and isinstance(left, Type)
+                ):
+                    relation_error(relation, left, right)
+            elif isinstance(relation, (Equal, NotEqual)):
+                # pylint: disable=too-many-boolean-expressions
+                if (
+                    (isinstance(left, Opaque) and not isinstance(right, (Opaque, Aggregate)))
+                    or (isinstance(left, Array) and not isinstance(right, (Array, Aggregate)))
+                    or (isinstance(left, Aggregate) and not isinstance(right, Composite))
+                ):
+                    relation_error(relation, left, right)
+                elif (
+                    (not isinstance(left, (Opaque, Aggregate)) and isinstance(right, Opaque))
+                    or (not isinstance(left, (Array, Aggregate)) and isinstance(right, Array))
+                    or (not isinstance(left, Composite) and isinstance(right, Aggregate))
+                ):
+                    relation_error(relation, left=right, right=left)
```

Comment:
The conditions seem to be the same just with `left` and `right` changed. I think it would make sense to introduce a function to reduce code duplication.

Reply:
Fixed in ee79d9c.